### PR TITLE
Assets in AT with poc+ (not to be merged right now)

### PR DESCRIPTION
**Pull request by jjos2372**
PR for us to discuss the upgrades being proposed before merging

### CODE REVIEW

File: html/ui/classic/js/brs.modals.account.js [hunk 1]
```diff
@@ -273,6 +273,12 @@ var BRS = (function(BRS, $, undefined) {
 			case 1:
 			    transactionType = "AT Payment";
 			    break;
+			case 2:
+				transactionType = "AT Asset Mold";
```
Comment (jjos2372) [inline html/ui/classic/js/brs.modals.account.js:277]: What is the "Mold" thing? Is it the "Issuance" or "AT Asset Issue"?

Comment (lao959) [reply html/ui/classic/js/brs.modals.account.js:277]: Yes, it is AT asset issuance.

File: src/brs/Asset.java [hunk 2]
```diff
@@ -10,6 +11,7 @@
   private final String name;
   private final String description;
   private final long quantityQNT;
+  private final long capabilityQNT;
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/Asset.java:14]: We need to think about that if the minable attribute for an asset issued by a normal address is reasonable, or will a user mint an asset from another user? what's the use case?

[the remark this answers is not in the document]
Comment (jjos2372) [reply src/brs/Asset.java:14]: Most platforms now allow you to create a token that are with fixed supply or not. But once set, it cannot be changed. One use case would be stable coins, pegged coins, etc. So every time you get a new deposit you mint more tokens. It is not trust-less, but it is how they do it.

File: src/brs/Attachment.java [hunk 3]
```diff
@@ -478,17 +478,17 @@ public Any getProtobufMessage() {
     }
 
     @Override
-      public TransactionType getTransactionType() {
-        return TransactionType.AutomatedTransactions.AT_PAYMENT;
-      }
+    public TransactionType getTransactionType() {
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/Attachment.java:481]: Got it.

File: src/brs/TransactionType.java [hunk 4]
```diff
@@ -71,8 +71,13 @@
 
   public static final byte SUBTYPE_AT_CREATION = 0;
   public static final byte SUBTYPE_AT_NXT_PAYMENT = 1;
+  private static final byte SUBTYPE_AT_ASSET_ISSUANCE = 2;
+  private static final byte SUBTYPE_AT_ASSET_TRANSFER = 3;
+
+  private static final byte SUBTYPE_ACCOUNT_CONTROL_EFFECTIVE_BALANCE_LEASING = 0;
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/TransactionType.java]: I will remove it.

File: src/brs/TransactionType.java [hunk 5]
```diff
@@ -963,8 +974,7 @@ void undoAttachmentUnconfirmed(Transaction transaction, Account senderAccount) {
       @Override
       void validateAttachment(Transaction transaction) throws BurstException.ValidationException {
         Attachment.ColoredCoinsAssetTransfer attachment = (Attachment.ColoredCoinsAssetTransfer)transaction.getAttachment();
-        if (transaction.getAmountNQT() != 0
-                || attachment.getComment() != null && attachment.getComment().length() > Constants.MAX_ASSET_TRANSFER_COMMENT_LENGTH
+        if (attachment.getComment() != null && attachment.getComment().length() > Constants.MAX_ASSET_TRANSFER_COMMENT_LENGTH
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/TransactionType.java:975]: Yes, removed the condition so we can generate the transfer asset truncation with BURST amount

File: src/brs/TransactionType.java [hunk 6]
```diff
@@ -2628,6 +2638,154 @@ public final boolean isSigned() {
...
 
+    public static final TransactionType AT_ASSET_ISSUANCE = new AutomatedTransactions() {
+      @Override
+      public final byte getSubtype() {
+        return TransactionType.SUBTYPE_AT_ASSET_ISSUANCE;
+      }
+
+      @Override
+      public String getDescription() {
+        return "AT Asset Mold";
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/TransactionType.java:2647]: Yes, but like to distinguish it from the asset issuance by a normal address

File: src/brs/TransactionType.java [hunk 7]
```diff
@@ -2628,6 +2638,154 @@ public final boolean isSigned() {
...
+    
+    public static final TransactionType AT_ASSET_TRANSFER = new AutomatedTransactions() {
+      @Override
+      public final byte getSubtype() {
+        return TransactionType.SUBTYPE_AT_ASSET_TRANSFER;
+      }
+
+      @Override
+      public String getDescription() {
+        return "AT Asset Mint";
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/TransactionType.java:2727]: Yes, Asset transfer from an AT and the asset is issued by the AT

File: src/brs/TransactionType.java [hunk 8]
```diff
@@ -2628,6 +2638,154 @@ public final boolean isSigned() {
...
+      @Override
+      void doValidateAttachment(Transaction transaction) throws BurstException.ValidationException {
+        Attachment.ATColoredCoinsAssetTransfer attachment = (Attachment.ATColoredCoinsAssetTransfer)transaction.getAttachment();
+        if (attachment.getComment() != null && attachment.getComment().length() > Constants.MAX_ASSET_TRANSFER_COMMENT_LENGTH
+                || attachment.getAssetId() == 0) {
+          throw new BurstException.NotValidException("Invalid asset transfer amount or comment: " + JSON.toJsonString(attachment.getJsonObject()));
+        }
+        if (transaction.getVersion() > 0 && attachment.getComment() != null) {
+          throw new BurstException.NotValidException("Asset transfer comments no longer allowed, use message " +
+                  "or encrypted message appendix instead");
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/TransactionType.java]: Fixed

File: src/brs/at/AT.java [hunk 9]
```diff
@@ -258,8 +259,44 @@ public void notify(Block block) {
                 accountService.addToBalanceAndUnconfirmedBalanceNQT(accountService.getAccount(AtApiHelper.getLong(atTransaction.getSenderId())), -atTransaction.getAmount());
                 accountService.addToBalanceAndUnconfirmedBalanceNQT(accountService.getOrAddAccount(AtApiHelper.getLong(atTransaction.getRecipientId())), atTransaction.getAmount());
 
-                Transaction.Builder builder = new Transaction.Builder((byte) 1, Genesis.getCreatorPublicKey(),
+                Transaction.Builder builder;
+                boolean messageIsText = false;
+                
+                if(atTransaction.getAsset() != null){
+
+                    //Issue Asset
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/at/AT.java:267]: Yes, I will DM you to see how to add the fee, I was trying to add but did not find a way.

File: src/brs/at/AT.java [hunk 10]
```diff
@@ -258,8 +259,44 @@ public void notify(Block block) {
...
+                        atTransaction.getAmount(), 0L, block.getTimestamp(), (short) 1440, atTransaction.getAsset());
+
+                    //set builder id to asset id in atTansaction, so the issued asset will have the same id
+                    builder.id(atTransaction.getAssetId());
+                    builder.fullHash(atTransaction.getFullHash());
+                }
+                else if(atTransaction.getAssetAmount() > 0) {
+
+                    //Transfer Asset
+                    messageIsText = true;
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/at/AT.java:285]: Give the user a friendly message after he/she mint some asset. like "Thank you for your support" in my FundsForFaucet sample contract.

File: src/brs/at/AT.java [hunk 11]
```diff
@@ -271,16 +308,24 @@ public void notify(Block block) {
 
                 byte[] message = atTransaction.getMessage();
                 if (message != null) {
-                    builder.message(new Appendix.Message(message, blockchain.getHeight()));
+                    if(messageIsText)
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/at/AT.java:311]: Give the user a friendly message after he/she mint some asset. like "Thank you for your support" in my FundsForFaucet sample contract.

File: src/brs/at/AtApi.java [hunk 12]
```diff
@@ -431,6 +431,30 @@
      */
     long getAmountForTxInA(AtMachineState state);
 
+    /**
+     * @param state AT machine state
+     * @return bool if A is a valid tx with @addr to tx asset id
+     */
+    long GetAssetIdForTxInA(AtMachineState state);
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/at/AtApi.java]: Fixed them

File: src/brs/at/AtApiPlatformImpl.java [hunk 13]
```diff
@@ -125,6 +131,119 @@ public long getAmountForTxInA(AtMachineState state) {
...
+        if (tx == null || (tx.getHeight() >= state.getHeight())) {
+            return -1;
+        }
+
+        ColoredCoinsAssetTransfer attachment = (tx.getAttachment() instanceof ColoredCoinsAssetTransfer)?(ColoredCoinsAssetTransfer)tx.getAttachment():null;
+        if (attachment != null) {
+            return attachment.getAssetId();
+        }
+
+        return 0;
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/at/AtApiPlatformImpl.java]: Fixed.

File: src/brs/schema/Db.java [hunk 14]
```diff
@@ -4,15 +4,41 @@
 package brs.schema;
 
 
-import brs.schema.tables.*;
+import java.util.ArrayList;
```
[the remark this answers is not in the document]
Comment (lao959) [reply src/brs/schema/Db.java:7]: These changes was done manually in order to compile the project with JDK 11, and all the changes were checked in in my first comment. I can exclude them in next pull request after we are done the reviews in this one.

File: src/brs/Attachment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (damccull) [reply src/brs/Attachment.java:481]: Why not provide a formatting guide based on Java's most popular formatting, then possibly provide formatting config files right in the repo that various IDEs will automatically pick up and use when autoformatting?

**Comment by jjos2372** [comment]
Implemented on another branch, thank you!

**Pull request closed without merging**